### ISSUE TITLE
Support for building on Apple Silicon Simulators

### ISSUE THREAD

**Issue by JonathanDowning** [braintree/braintree_ios#564]
**Integration Details (please complete the following information):**
 - SDK/Library Version: 5.0.0-beta1
 - Environment: Sandbox
 - iOS Version: iOS 14.3
 - Device iOS Simulator on Apple Silicon
 - Integration type & version Swift Package Manager

**Describe the bug**
When using Braintree integrated with Swift Package Manager, on an Apple Silicon Mac, I am unable to compile the framework due to MagnesSDK being incompatible with this configuration. 

**To Reproduce**
Compile a project utilizing the Braintree 5.0.0-beta SDK using an Apple Silicon Mac 

**Expected behavior**
A successful build of the project

**Screenshots**
<img width="1404" alt="Screen Shot 2020-12-18 at 05 34 06" src="https://user-images.githubusercontent.com/2230175/102604767-a934b880-40f2-11eb-9ec7-8ded7982e0ba.png">

**Comment by scannillo** [comment]
Hello @JonathanDowning - we don't currently have an Apple Silicon Mac to replicate this on to troubleshoot, however we will reach out to our Magnes SDK team about this issue.

Just to confirm that you have the Braintree modules properly included via SPM, can you confirm which Braintree modules you have in your project settings under General > Frameworks, Libraries, and Embedded Content (and that they are each set to Embed & Sign)?

**Comment by JonathanDowning** [comment]
Hey @scannillo and Happy New Year!

I've been playing around with this issue this morning and here are some observations:

1. The project is using `braintree-ios` version `5.0.0-beta1`
2. `BraintreePayPal` and `PayPalDataCollector` have been added to the project's Frameworks, Libraries, and Embedded Content
3. The project successfully builds for `Any iOS Device (arm64)`
4. The project successfully builds for iOS 13 and iOS 12 simulators
5. The project fails to build for iOS 14 simulators, emitting the following error: `Cannot find 'MagnesSDK' in scope`

Thoughts:
I suspect MagnesSDK need to be built with an architecture slice appropriate for iOS Simulators running natively in Apple Silicon mode.
iOS 12 and 13 simulators currently [run in Rosetta mode](https://steipete.com/posts/apple-silicon-mac-mini-for-ci/) which may explain why these don't emit the error, whereas iOS 14 sims do.

Having inspected Firebase for iOS's binary dependencies, which do not exhibit this issue, I discovered their xcframeworks have an `arm64` slice in the simulator variants.
It appears `PPRiskMagnes.xcframework` is missing this `arm64` slice and thus cannot compile for arm64 (Apple Silicon) simulators.

<img width="1368" alt="Screen Shot 2021-01-04 at 07 18 05" src="https://user-images.githubusercontent.com/2230175/103534702-63e7e900-4e5d-11eb-817d-054083e28dfd.png">

**Comment by sestevens** [comment]
Hi @JonathanDowning, thanks for providing all of this detail! I just opened a PR updating to a newer version of the Magnes SDK that includes an arm64 simulator slice. We'll post an update to this issue when this change is released. Feel free to check out the PR branch if you like!

**Comment by JonathanDowning** [comment]
@sestevens thank you for the prompt response, I've tested it and I can now use iOS 14 simulators on Apple Silicon, yay!

**Comment by sestevens** [comment]
@JonathanDowning Awesome, thanks for letting us know! 👏

**Comment by sestevens** [comment]
This change has been released in 5.0.0-beta2.

**Comment by ejensen** [comment]
The same issue with not being able to use the iOS 14 simulators still occurs with the `PPRiskMagnes` and `CardinalMobile` frameworks
<img width="1053" alt="Screen Shot 2021-01-26 at 12 23 50 PM" src="https://user-images.githubusercontent.com/915431/105901498-50e6b580-5fd2-11eb-8776-c3d82ce31767.png">

**Comment by scannillo** [comment]
👋  Hi @ejensen. That looks like a different issue than the one being reported in this thread. In your build settings, can you set `Excluded Architectures` to add `Any iOS Simulator SDK` with a value of `arm64` for both Debug and Release mode ([see related issue](https://developer.apple.com/forums/thread/656509))? If you would like to follow up about this - please open a new GitHub issue.

**Comment by ejensen** [comment]
@scannillo excluding the `arm64` architecture won't fix this issue where the iOS 14 simulator on Apple Silicon requires the `arm64` architecture slice that `KountDataCollector` and `CardinalMobile` don't provide. The error in my previous post's screenshot shows that Xcode is finding the `Simulator-x86_64` architecture in those frameworks and not the `arm64` slice it needs to run.

Similar to [the screenshot @JonathanDowning posted above](https://github.com/braintree/braintree_ios/issues/564#issuecomment-753944946), `KountDataCollector ` and `CardinalMobile` don't have the `arch64` slice which is now included in the 5.10.0 `PPRiskMagnes` framework.
![archs](https://user-images.githubusercontent.com/915431/106029933-7a5f1a00-6082-11eb-8fd8-923093f69d9a.png)

The simulators won't run the `Simulator-x86_64` code unless it's forced to run with Rosetta 2 ([workaround mentioned above is to use the iOS 13 simulator](https://github.com/braintree/braintree_ios/issues/564#issuecomment-753944946)
I'm hoping the `arm64` can be included in the `PPRiskMagnes` and `CardinalMobile` frameworks we can run the iOS 14 simulator on Apple Silicon

**Comment by scannillo** [comment]
Ok - thank you for the additional detail @ejensen. This makes sense, and we are going to reach out to our partners at `CardinalMobile` and `Kount` ASAP.

Can you confirm though that you're able to build OK on a real device?

**Comment by ejensen** [comment]
@scannillo Thanks! Yes, I'm able to run on a real device. This issue only surfaces on the Simulator

**Comment by sestevens** [comment]
This has been fixed for `PPRiskMagnes` and `KountDataCollector` in [version 5.0.0](https://github.com/braintree/braintree_ios/releases/tag/5.0.0). We're waiting for a fix for `CardinalMobile`, so we'll leave this issue open for now.

Please see our [migration guide](https://github.com/braintree/braintree_ios/blob/5.0.0/V5_MIGRATION.md) for steps to upgrade your integration.

**Comment by JackoPlane** [comment]
Hi @sestevens, Do you have any sort of timeline for supporting M1 macs with `CardinalMobile` ?

**Comment by scannillo** [comment]
Hi @JackoPlane - our Fraud Tools team is regularly bringing up this request with CardinalMobile, however they haven't given us information about arm64 support yet. We will keep you posted as soon as we have more information and will continue to raise this as a pressing issue to them.

**Comment by alexito4** [comment]
@scannillo do you have any update on this? Currently this is the only library that is blocking us in M1 which is causing trouble in our development team.

**Comment by lordukr** [comment]
Hey Braintree team, any updates regarding this?

**Comment by scannillo** [comment]
Hi everyone - thank you for your activity on this issue. We want to get this resolved as quickly as possible, but are still blocked by our MPI (3DS) provider `CardinalMobile`. We are actively communicating this concern to them on a weekly basis but they still haven't given us an ETA. We will keep pushing on this as much as we can, and keep you posted.

**Comment by alexito4** [comment]
@scannillo if you need more arguments to push for be aware that it seems that current workarounds won't work anymore https://developer.apple.com/forums/thread/681749 so this will literally block all your customers pretty soon.

**Comment by lordukr** [comment]
@scannillo nothing new?

**Comment by scannillo** [comment]
@lordukr - no news, unfortunately. Will keep on it and keep you posted.

**Comment by bofeizhu** [comment]
@scannillo It has been 7 months since this issue was opened. Maybe it's time for your team to switch to another provider that can offer more reliable support.

**Comment by scannillo** [comment]
> @scannillo It has been 7 months since this issue was opened. Maybe it's time for your team to switch to another provider that can offer more reliable support.

Hi @bofeizhu. I agree in that this current lack of support from CardinalMobile is unacceptable. My team and I are actively and frequently communicating these concerns to our leaders. In the meantime, we will continue to prod Cardinal for this update.

**Comment by Onno113** [comment]
Hello, I am just writing to mention that this is also problem for the company I am working for (FREE NOW) because we have the first computer for CI/CD with Apple Silicon architecture. The Braintree SDK is the only reason we cannot use it, it seems.

**Comment by JackoPlane** [comment]
Echoing @dweiner13 comments, 

@scannillo Could we have an update on where things are at and how the community can assist in applying pressure to get support for Apple Silicon? Is this something we should be brining up with our account manager?

**Comment by scannillo** [comment]
> @scannillo Could we have an update on where things are at and how the community can assist in applying pressure to get support for Apple Silicon? Is this something we should be brining up with our account manager?

Hi everyone. Cardinal has still given us no update. We are applying as much pressure as we can on this.

It would definitely help if you raise this concern to your account manager! Thank you!

**Comment by DagAgren** [comment]
@scannillo The rumours are that running the simulator under Rosetta no longer works on macOS 12, which would block all development of any app that uses Cardinal. So this is getting quite urgent to fix, as macOS 12 is coming fairly soon.

See for instance https://developer.apple.com/forums/thread/681749.

**Comment by skywalkerlw** [comment]
Can not wait !!!

**Comment by alexjameslittle** [comment]
We at Depop are also blocked by this issue

**Comment by scannillo** [comment]
Hi everyone. Just an update for you all - we've been consistently raising this issue to Cardinal, and they have given us an ETA of making this update in late October. We will do everything we can to push that they deliver on this timeframe estimate.

**Comment by scannillo** [comment]
Hi @shepting - we have this[ PR open ](https://github.com/braintree/braintree_ios/pull/750)adding support for M1 👏 . Will keep this thread posted when we cut the release. Should be soon, assuming our CI & testing succeeds as expected.

**Comment by scannillo** [comment]
Hello everyone - **Braintree iOS v5.5.0** is released. It allows all SPM, CocoaPods, and Carthage users on `xcframeworks` to get Apple Silicon simulator support. Please see the [release notes](https://github.com/braintree/braintree_ios/releases/tag/5.5.0) for the caveat for Carthage users on `framework`s. Another Cardinal bug caused this limitation, but most of you should be unblocked.

Please leave feedback if the SDK works as expected for you!

**Comment by bofeizhu** [comment]
@scannillo Could you release `BraintreeDropIn` for arm64-sim as well (drop exclude archs)

**Comment by scannillo** [comment]
Hi @bofeizhu - definitely. It is in PR now (https://github.com/braintree/braintree-ios-drop-in/pull/357). Once it's approved internally, we'll get a release out ASAP. But in the meantime feel free to use that branch `arm-64-support` if you want.

**Comment by scannillo** [comment]
I am going to close this issue now that v5.5.0 is released with arm64 support 👏 . Feel free to open a new issue if any issues arise! Thank you all for your patience and for commenting on this thread, it helps us a ton!
